### ISSUE TITLE
Parse error on `$do`

### ISSUE THREAD

**Issue by sjakobi** [ndmitchell/hlint#1222]
See https://github.com/haskell-CI/haskell-ci/runs/2108280754.

It's valid Haskell isn't it?!

**Comment by RyanGlScott** [comment]
It is valid Haskell, but it may not be valid GHC. Here is a minimal example of this phenomenon:

```hs
module Foo where

f :: ()
f = id $do ()
```

This will parse if compiled without any language extensions, but it will fail to parse if compiled with `TemplateHaskell` enabled. This is because GHC will try parse `$do` as a Template Haskell splice but fail to do so.

Note that unlike #1212, this is not something that is new in GHC 9.0. Even pre-9.0 versions of GHC will fail to parse the program above with `TemplateHaskell` enabled.

**Comment by phadej** [comment]
This is a regression:

```
% hlint Bar.hs 
Bar.hs:4:5-10: Warning: Redundant id
Found:
  id $do
Perhaps:
  $do

1 hint

% hlint --version
HLint v3.2.6, (C) Neil Mitchell 2006-2020
```

**Comment by ndmitchell** [comment]
If you run hlint with `-XNoTemplateHaskell` it should still continue to parse this code. Although this particular fragment is decidedly odd, and breaks if someone sets `TemplateHaskell` in their Cabal file or similar. Is there any reason that `$do` is the right thing to write, and that a space before the `do` wouldn't be a better idea?

**Comment by sjakobi** [comment]
I didn't mind getting a "style hint" about a missing space, although the error is somewhat surprising.

I wouldn't mind if this is closed as "won't fix".

**Comment by phadej** [comment]
I think enabling `-XTemplateHaskell` by default is a bad idea. (Or generally any syntax stealing extensions). Especially as not all of them are enabled:

```
[polinukli] /codetmp % hlint Bar.hs                                              
Bar.hs:5:5-6: Error: Parse error: on input `<-'
Perhaps you intended to use RecursiveDo
Found:
    f :: ()
    f = id $ mdo
  >   x <- x
      return x
  

1 hint

% hlint --version
HLint v3.3, (C) Neil Mitchell 2006-2021
```

And in that case parser is helpful in suggesting to enable `RecursiveDo`,
yet why `$do` is invalid beyond people not aware of Template Haskell.
```

**Comment by ndmitchell** [comment]
The policy has always been to enable extensions that steal some syntax, if they are common. There are people who put their extensions in files that HLint can't see, and while you could tell everyone to properly configure HLint, that's hostile to beginners. Template Haskell is common and people generally shouldn't write `$do`. If Ed went on twitter and started telling everyone to write `$do` for some library, we'd probably turn off TemplateHaskell. I think the reason we disable recursive do is because it steals the name `rec`, which is somewhat common in Haskell - if it was just the `mdo` we'd have it enabled.

**Comment by phadej** [comment]
All projects I know which use HLint have `.hlint.yaml`. Is there a way to reset which extensions are enabled. I.e. specify which are enabled and not-via delta to detault HLint set?

**Comment by phadej** [comment]
Perfectly things like default extensions should also be documented. At first I attribute the parsing error to whitespace handling changes in lexer in GHC-9.0 (which made me worry whether if it's broken badly). Template Haskell **was not** the first idea why things were broken.

**Comment by googleson78** [comment]
You can indeed specify which extensions are enabled with the `.hlint.yaml` file - see [this section](https://github.com/ndmitchell/hlint#language-extensions)

**Comment by ndmitchell** [comment]
If you want more default Haskell, `-XHaskell2010` as the first extension should reset the list to that.

**Comment by phadej** [comment]
@ndmitchell nope.

```
% hlint -XHaskell2010 Foo2.hs
Foo2.hs:4:9-10: Error: Parse error: on input `do'
Found:
    f :: ()
  > f = id $do ()
  

1 hint
```

**Comment by ndmitchell** [comment]
@shayne-fletcher is fixing this as #1233

**Comment by ndmitchell** [comment]
FWIW, it is documented at https://github.com/ndmitchell/hlint#bugs-and-limitations:

> foo $bar means something different with TemplateHaskell. These extensions can be disabled with -XNoMagicHash or -XNoTemplateHaskell etc

It's a shame it breaks some things though - I think fixing the `-XHaskell2010` would provide a good experience to fixing that.
